src/hotspot/share/oops/constantPool.cpp

```diff
@@ -2443,3 +2334,85 @@ void ConstantPool::verify_on(outputStream* st) {
     guarantee(pool_holder()->is_klass(),    "should be klass");
   }
 }
+
+void BSMAttributeEntries::deallocate_contents(ClassLoaderData* loader_data) {
+  MetadataFactory::free_array<u4>(loader_data, this->_offsets);
+  MetadataFactory::free_array<u2>(loader_data, this->_bootstrap_methods);
+  this->_offsets = nullptr;
+  this->_bootstrap_methods = nullptr;
+}
+
+void BSMAttributeEntries::copy_into(InsertionIterator& iter, int num_entries) const {
+  assert(num_entries + iter._cur_offset <= iter.insert_into->_offsets->length(), "must");
+  for (int i = 0; i < num_entries; i++) {
+    const BSMAttributeEntry* bsmae = entry(i);
+    BSMAttributeEntry* bsmae_new = iter.reserve_new_entry(bsmae->bootstrap_method_index(), bsmae->argument_count());
+    bsmae->copy_args_into(bsmae_new);
+  }
+}
+
+BSMAttributeEntries::InsertionIterator BSMAttributeEntries::start_extension(const BSMAttributeEntries& other, ClassLoaderData* loader_data, TRAPS) {
+  return start_extension(other.number_of_entries(), other.array_length(), loader_data, CHECK_(BSMAttributeEntries::InsertionIterator()));
+}
+
+BSMAttributeEntries::InsertionIterator
+BSMAttributeEntries::start_extension(int number_of_entries, int array_length,
+                                   ClassLoaderData* loader_data, TRAPS) {
+  InsertionIterator extension_iterator(this, this->number_of_entries(), this->array_length());
+  int new_number_of_entries = this->number_of_entries() + number_of_entries;
+  int new_array_length = this->array_length() + array_length;
+  int invalid_index = new_array_length;
+
+  Array<u4>* new_offsets =
+    MetadataFactory::new_array<u4>(loader_data, new_number_of_entries, invalid_index, CHECK_(InsertionIterator()));
+  Array<u2>* new_array = MetadataFactory::new_array<u2>(loader_data, new_array_length, CHECK_(InsertionIterator()));
+  { // Copy over all the old BSMAEntry's and their respective offsets
+    BSMAttributeEntries carrier(new_offsets, new_array);
+    InsertionIterator copy_iter(&carrier, 0, 0);
+    copy_into(copy_iter, this->number_of_entries());
+  }
+  // Replace content
+  deallocate_contents(loader_data);
+  _offsets = new_offsets;
+  _bootstrap_methods = new_array;
+  return extension_iterator;
+}
+
+
+void BSMAttributeEntries::append(const BSMAttributeEntries& other, ClassLoaderData* loader_data, TRAPS) {
+  if (other.number_of_entries() == 0) {
+    return; // Done!
+  }
+  InsertionIterator iter = start_extension(other, loader_data, CHECK);
+  other.copy_into(iter, other.number_of_entries());
+  end_extension(iter, loader_data, CHECK);
+
+}
+
+void BSMAttributeEntries::end_extension(InsertionIterator& iter, ClassLoaderData* loader_data,
+                                      TRAPS) {
```

Comment:
```suggestion
void BSMAttributeEntries::end_extension(InsertionIterator& iter, ClassLoaderData* loader_data, TRAPS) {
```
This line is still shorter than 2382 above